pkgs/native_assets_cli/lib/src/config.dart

```diff
@@ -0,0 +1,538 @@
+      // directory and then call [BuildConfigBuilder.setupBuildRunConfig] &
+      // [LinkConfigBuilder.setupLinkRunConfig].
+      throw StateError('The checksum should be generated before setting '
+          'up the run configuration');
+    }
+    final hash = sha256
+        .convert(const JsonEncoder().fuse(const Utf8Encoder()).convert(json))
+        .toString()
+        // 256 bit hashes lead to 64 hex character strings.
+        // To avoid overflowing file paths limits, only use 32.
+        // Using 16 hex characters would also be unlikely to have collisions.
+        .substring(0, 32);
+    return hash;
+  }
+  //String computeChecksum() {
+  //final savedOutDir = json.remove(_outDirConfigKey);
+  //final savedSharedOutDir = json.remove(_outDirSharedConfigKey);
+  //final hash = sha256
+  //.convert(const JsonEncoder().fuse(const Utf8Encoder()).convert(json))
+  //.toString()
+  //// 256 bit hashes lead to 64 hex character strings.
+  //// To avoid overflowing file paths limits, only use 32.
+  //// Using 16 hex characters would also be unlikely to have collisions.
+  //.substring(0, 32);
+  //if (savedOutDir != null) json[_outDirConfigKey] = savedOutDir;
+  //if (savedSharedOutDir != null) {
+  //json[_outDirSharedConfigKey] = savedSharedOutDir;
+  //}
+  //return hash;
+  //}
+}
+
+const _targetOSConfigKey = 'target_os';
+const _buildModeConfigKey = 'build_mode';
+const _metadataConfigKey = 'metadata';
+const _outDirConfigKey = 'out_dir';
+const _outDirSharedConfigKey = 'out_dir_shared';
+const _packageNameConfigKey = 'package_name';
+const _packageRootConfigKey = 'package_root';
+const _supportedAssetTypesKey = 'supported_asset_types';
+
+final class BuildConfig extends HookConfig {
+  /// Whether this run is a dry-run, which doesn't build anything.
+  ///
+  /// A dry-run only reports information about which assets a build would
+  /// create, but doesn't actually create files.
+  final bool dryRun;
+
+  final bool linkingEnabled;
+
+  final Map<String, Metadata> metadata;
+
+  BuildConfig(super.json)
+      : dryRun = json.getOptional<bool>(_dryRunConfigKey) ?? false,
+        linkingEnabled = json.get<bool>(_linkingEnabledKey),
+        metadata = {
+          for (final entry
+              in (json.optionalMap(_dependencyMetadataKey) ?? {}).entries)
+            entry.key: Metadata.fromJson(as<Map<String, Object?>>(entry.value)),
+        };
+
+  Object? metadatum(String packageName, String key) =>
+      metadata[packageName]?.metadata[key];
+}
+
+final class BuildConfigBuilder extends HookConfigBuilder {
+  void setupBuildConfig({
+    required bool dryRun,
+    required bool linkingEnabled,
+    Map<String, Metadata>? metadata,
+  }) {
+    json[_dryRunConfigKey] = dryRun;
+    json[_linkingEnabledKey] = linkingEnabled;
+    if (metadata != null) {
+      json[_dependencyMetadataKey] = {
+        for (final key in metadata.keys) key: metadata[key]!.toJson(),
+      };
+    }
+  }
+
+  void setupBuildRunConfig({
```

Comment:
> I kind of like this splitting up in separate parts.

:+1: it groups related things together. Much better then the enormous parameter lists of unrelated things.

> However, it leads to the possiblity where an embedder might forget to call one of the parts.

We only have 2 embedders and they have to call that in 2 places (build & link). I'm not concerned we will forget about it (we'd know because each embedder is bundling things differently, so each embedder will have to have end-to-end tests that would fail if anything doesn't work - so it would be discovered before landing any code).

> Of course, having the 'subtypes' calling the 'parent' ones leads to forwarding a million parameters, which is also ugly.

This isn't really possible, there's no such things as "subtypes", no hierarchy. A bundling tool can cherry-pick individual pieces it wants to support. Imagine later on in flutter it looks like this
```dart
BuildConfigBuilder()
  ..setupCodeConfig(...)
  ..setupFontConfig(...)
  ..setupIconConfig(...)
  ..setupShaderConfig(...)
```
none of them is a subtype of the other - they form no hierarchy.

Reply:
As we discussed, the PR wasn't quite ready. I've added now a `configValidator` that verifies the soundness of the final configuration.